--- linux_fdp/drivers/nvme/host/nvme.h	2025-02-12 15:09:24.296275035 +0000
+++ linux_fdp/drivers/nvme/host/nvme_fdp.h	2025-02-12 15:08:37.668041491 +0000
@@ -446,6 +446,7 @@
 	u8	csi;
 };
 
+#define NVME_MAX_PLIDS   (WRITE_LIFE_EXTREME + 1)
 
 /*
  * Anchor structure for namespaces.  There is one for each namespace in a
@@ -464,6 +465,8 @@
 	bool			shared;
 	bool			passthru_err_log_enabled;
 	int			instance;
+	u16			nr_plids;
+	u16			plids[NVME_MAX_PLIDS];
 	struct nvme_effects_log *effects;
 	u64			nuse;
 	unsigned		ns_id;
